tests/topotests/lib/common_config.py

```diff
@@ -1619,60 +1625,60 @@ def interface_status(tgen, topo, input_dict):
+    * `expected`: if False then the return logic is inverted, except for exceptions,
+                      (i.e., a False or errmsg (str) function return ends the retry loop,
+                      and returns that False or str value)
     """
 
     def _retry(func):
         @wraps(func)
         def func_retry(*args, **kwargs):
-            _wait = kwargs.pop("wait", wait)
-            _attempts = kwargs.pop("attempts", attempts)
-            _attempts = int(_attempts)
-            if _attempts < 0:
-                raise ValueError("attempts must be 0 or greater")
+            _retry_sleep = 2
+
+            _retry_timeout = kwargs.pop("retry_timeout", retry_timeout)
+            _expected = kwargs.pop("expected", expected)
+            _initial_wait = kwargs.pop("initial_wait", initial_wait)
+
+            retry_until = datetime.now() + timedelta(seconds=_retry_timeout + _initial_wait)
 
             if initial_wait > 0:
                 logger.info("Waiting for [%s]s as initial delay", initial_wait)
                 sleep(initial_wait)
 
-            _return_is_str = kwargs.pop("return_is_str", return_is_str)
-            _return_is_dict = kwargs.pop("return_is_str", return_is_dict)
-            _expected = kwargs.setdefault("expected", True)
-            kwargs.pop("expected")
-            for i in range(1, _attempts + 1):
+            invert_logic = not _expected
+            attempts = 0
+            while True:
+                seconds_left = (retry_until - datetime.now()).total_seconds()
+                attempts += 1
                 try:
                     ret = func(*args, **kwargs)
                     logger.debug("Function returned %s", ret)
-                    if _return_is_str and isinstance(ret, bool) and _expected:
-                        return ret
-                    if (
-                        isinstance(ret, str) or isinstance(ret, unicode)
-                    ) and _expected is False:
-                        return ret
-                    if _return_is_dict and isinstance(ret, dict):
-                        return ret
 
-                    if _attempts == i:
-                        generate_support_bundle()
+                    negative_result = ret is False or is_string(ret)
+                    if negative_result == invert_logic:
                         return ret
+
                 except Exception as err:
-                    if _attempts == i:
-                        generate_support_bundle()
-                        logger.info("Max number of attempts (%r) reached", _attempts)
+                    logger.info("Function raised exception: %s", str(err))
+                    if seconds_left < 0:
                         raise
-                    else:
-                        logger.info("Function returned %s", err)
-                if i < _attempts:
-                    logger.info("Retry [#%r] after sleeping for %ss" % (i, _wait))
-                    sleep(_wait)
+                finally:
```

Comment:
Just an input:
"The finally block always executes after normal termination of try block or after try block terminates due to some exception."

We have put same if condition(if seconds_left <= 0) 2 times, either we can omit finally or remove one if condition. Please check once.

Reply:
I don't follow this comment, the use of the finally block has consolidate the the 2 copies of code into a single one.